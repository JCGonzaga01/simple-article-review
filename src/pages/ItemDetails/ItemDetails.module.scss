@import "helpers/stylings/mixins";
@import "helpers/stylings/global";

.wrapper {
  @include align-center;
  @media (max-width: 1569px) {
    flex-direction: column;
  }
  margin-bottom: 100px;

  .imageSrc {
    @include media-query(sp) {
      background-size: cover;
      width: 100%;
    }

    @media (max-width: 1569px) {
      margin-right: 0;
      margin-bottom: 100px;
    }

    margin-right: 131px;
    width: 682px;
    height: 1107px;
    background-repeat: no-repeat;

    .tag {
      background: $tagBg;
      color: $tag;
      display: inline-table;
      font-family: $fontType1;
      font-size: $fontM;
      font-weight: 400;
      letter-spacing: 2px;
      line-height: $lineheightM;
      padding: 15px 0 0 15px;
      text-align: right;
      transform: rotate(180deg);
      writing-mode: vertical-lr;
    }
  }

  .contentContainer {
    width: 564px;

    & > div:last-child > div:last-child {
      margin-bottom: 0;
    }

    .title {
      color: $title;
      font-family: $fontType1;
      font-size: $fontTitle;
      font-weight: 700;
      letter-spacing: -2px;
      line-height: $lineheightTitle;
      margin-bottom: 40px;
      width: 445px;
    }

    .sort {
      @include align-center;
      justify-content: flex-start;
      color: $link;
      cursor: pointer;
      font-family: $fontType1;
      font-size: $fontXS2;
      font-weight: 700;
      letter-spacing: 0px;
      line-height: $lineheightXS2;
      margin-bottom: 30px;
      text-transform: uppercase;

      & > span {
        margin-right: 6px;
      }
    }

    .questionTitle {
      color: $black;
      font-family: $fontType1;
      font-size: $fontXL;
      font-weight: 700;
      letter-spacing: -1px;
      line-height: $lineheightXL;
      margin-bottom: 19px;
    }

    .questionDate {
      @include align-center;
      color: $date;
      justify-content: flex-start;
      font-family: $fontType1;
      font-size: $fontXS2;
      font-weight: 700;
      letter-spacing: 0px;
      line-height: $lineheightXS2;
      margin-bottom: 11px;

      .line {
        background: $date;
        border: none;
        height: 2px;
        margin: 0 16px 0 0;
        width: 32px;
      }
    }

    .questionText {
      color: $content;
      font-family: $fontType2;
      font-size: $fontS;
      font-weight: 300;
      line-height: $lineheightM;
      margin-bottom: 42px;
    }
  }
}
