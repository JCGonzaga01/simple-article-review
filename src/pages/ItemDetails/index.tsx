import React, { useState, useEffect } from "react";
import { useLocation } from "react-router-dom";
import moment from "moment";
import queryString from "query-string";
import data from "_root/data.json";
import styles from "./ItemDetails.module.scss";

const ItemDetails: React.FC = () => {
  const qs = queryString.parse(useLocation().search);
  const details = data.find((item) => JSON.stringify(item.id) === qs.id);
  const [questions, setQuestions] = useState(details?.questions || []);
  const [sortText, setSortText] = useState("sort by latest");
  const url = process.env.PUBLIC_URL + `/assets/itemdetails/${details?.image}`;

  useEffect(() => {
    return () => setQuestions(details?.questions || []);
  }, [details]);

  const handleOnClickSort = () => {
    const isSortLatest = sortText === "sort by latest";
    const sortedQuestions =
      [...questions].sort((a, b) => {
        const curr = moment(a.date).utc().format("YYYYMMDD") as any;
        const next = moment(b.date).utc().format("YYYYMMDD") as any;
        return isSortLatest ? next - curr : curr - next;
      }) || [];
    setQuestions([...sortedQuestions]);
    setSortText(isSortLatest ? "sort by earliest" : "sort by latest");
  };

  return !details ? (
    <></>
  ) : (
    <div className={styles.wrapper}>
      <div className={styles.imageSrc} style={{ backgroundImage: `url(${url})` }}>
        <span className={styles.tag}>{details.tag}</span>
      </div>
      <div className={styles.contentContainer}>
        <div className={styles.title}>{details.title}</div>
        <div className={styles.sort} onClick={handleOnClickSort}>
          <span>{sortText}</span>
          <img src={process.env.PUBLIC_URL + "/assets/sort.png"} alt={"sort"} />
        </div>
        {questions.map((item, idx) => {
          return (
            <div key={`${details.id}:${idx}`}>
              <div className={styles.questionTitle}>{item.title}</div>
              <div className={styles.questionDate}>
                <hr className={styles.line} />
                {moment(item.date).utc().format("DD/MM/YYYY")}
              </div>
              <div className={styles.questionText}>{item.text}</div>
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default ItemDetails;
