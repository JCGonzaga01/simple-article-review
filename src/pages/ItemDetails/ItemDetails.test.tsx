import React from "react";
import renderer from "react-test-renderer";
import ItemDetails from "./index";

jest.mock("react-router-dom", () => ({
  useLocation: () => ({
    search: "id=1001",
  }),
}));

describe("<ItemDetails />", () => {
  let wrapper: any;
  let instance: any;

  beforeEach(() => {
    wrapper = renderer.create(<ItemDetails />);
    instance = wrapper.root;
  });

  test("should mount component", () => {
    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("should render details", () => {
    expect(instance.findByProps({ className: "wrapper" })).toBeTruthy();
  });

  test("should render questions", () => {
    expect(instance.findByProps({ className: "contentContainer" }).children).toHaveLength(7);
  });

  test("should handle sort", () => {
    let sortText = instance.findByProps({ className: "sort" }).children[0]._fiber.stateNode.props
      .children;
    expect(sortText).toEqual("sort by latest");

    const sortButton = instance.find((element: any) => element.props.className === "sort");
    sortButton.props.onClick();

    sortText = instance.findByProps({ className: "sort" }).children[0]._fiber.stateNode.props
      .children;
    expect(sortText).toEqual("sort by earliest");
  });
});
