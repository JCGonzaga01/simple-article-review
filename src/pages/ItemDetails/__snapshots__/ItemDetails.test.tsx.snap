// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemDetails /> should mount component 1`] = `
<div
  className="wrapper"
>
  <div
    className="imageSrc"
    style={
      Object {
        "backgroundImage": "url(/assets/itemdetails/1001.jpg)",
      }
    }
  >
    <span
      className="tag"
    >
      Toga Group
    </span>
  </div>
  <div
    className="contentContainer"
  >
    <div
      className="title"
    >
      Lorem ipsum dolor sit amet
    </div>
    <div
      className="sort"
      onClick={[Function]}
    >
      <span>
        sort by latest
      </span>
      <img
        alt="sort"
        src="/assets/sort.png"
      />
    </div>
    <div>
      <div
        className="questionTitle"
      >
        Content title 1
      </div>
      <div
        className="questionDate"
      >
        <hr
          className="line"
        />
        11/05/1984
      </div>
      <div
        className="questionText"
      >
        eiusmod sint esse est velit et nostrud quis mollit elit nostrud est proident reprehenderit ea ut fugiat sunt aliqua consectetur dolor irure ut aliqua voluptate
      </div>
    </div>
    <div>
      <div
        className="questionTitle"
      >
        Content title 2
      </div>
      <div
        className="questionDate"
      >
        <hr
          className="line"
        />
        19/04/1996
      </div>
      <div
        className="questionText"
      >
        ut in nisi qui nulla voluptate minim sunt elit ad id commodo pariatur ipsum veniam in pariatur cillum do elit incididunt enim do excepteur dolor
      </div>
    </div>
    <div>
      <div
        className="questionTitle"
      >
        Content title 3
      </div>
      <div
        className="questionDate"
      >
        <hr
          className="line"
        />
        09/12/2004
      </div>
      <div
        className="questionText"
      >
        eiusmod officia est velit tempor voluptate in in non est non dolore cillum eiusmod adipisicing laboris minim fugiat adipisicing voluptate quis nulla irure anim excepteur
      </div>
    </div>
    <div>
      <div
        className="questionTitle"
      >
        Content title 4
      </div>
      <div
        className="questionDate"
      >
        <hr
          className="line"
        />
        03/11/1970
      </div>
      <div
        className="questionText"
      >
        amet cillum veniam dolor duis incididunt Lorem cupidatat Lorem officia quis consectetur dolor non laborum occaecat pariatur est nulla aliqua aliqua ex veniam dolore laborum
      </div>
    </div>
    <div>
      <div
        className="questionTitle"
      >
        Content title 5
      </div>
      <div
        className="questionDate"
      >
        <hr
          className="line"
        />
        30/03/1976
      </div>
      <div
        className="questionText"
      >
        sunt eiusmod cupidatat dolore nostrud labore fugiat enim do eiusmod dolore cupidatat eu dolore deserunt aliquip sint laborum consectetur voluptate irure ea veniam exercitation deserunt
      </div>
    </div>
  </div>
</div>
`;
